### PR TITLE
Add Player HUD & UI Manager Script (Player HUD #34)

### PR DESCRIPTION
**Pull request by LuxPoseidon**
Updated Beta HUD to include heath, stamina. food, water, and money statistics.

Health and stamina work natively (vanilla). Food and water functionality are pending a survival stats script, similarly money is pending money integration.

Add
+ EL_BetaHud.c

Modify
+ Beta.Layout

ui

### CODE REVIEW

File: src/Scripts/Game/UI/EL_BetaHud.c [hunk 1]
```diff
@@ -1,3 +1,120 @@
...
+		{
+			Print("no player found");
+			return;
+		}
+		
+		playerController = SCR_CharacterControllerComponent.Cast(player.FindComponent(SCR_CharacterControllerComponent));
+		if (playerController) 
+		{
+			Print("Found character controller component")
+		}
```
Comment (Thurston00) [inline src/Scripts/Game/UI/EL_BetaHud.c:85]: This entire statement isn't needed.

File: src/Scripts/Game/UI/EL_BetaHud.c [hunk 2]
```diff
@@ -1,3 +1,120 @@
...
+	//UpdateValues needs to be called upon respawning as to reset the UI for healthSlider so it isnt stuck at 0 until you take damage -- KNOWN BUG
+	override event void UpdateValues(IEntity owner, float timeSlice)
+	{
+		if (!playerController)
+		{
+			IEntity player = SCR_PlayerController.GetLocalControlledEntity();
+			if (!player) 
+			{
+				return;
+			}
```
Comment (Thurston00) [inline src/Scripts/Game/UI/EL_BetaHud.c:99]: You're using multiple types of different formats here for `if (!foo) return;`, the general way to go for this is how you've done it on line 101.

File: src/Scripts/Game/UI/EL_BetaHud.c [hunk 3]
```diff
@@ -1,3 +1,120 @@
...
+			
+			DMC = DamageManagerComponent.Cast(player.FindComponent(DamageManagerComponent));
+			if (!DMC) return;
+			
+			//add things here to find component for survivalstats
+			//SSM = EL_SurvivalStatsManagerComponent.Cast(player.FindComponent(EL_SurvivalStatsManagerComponent));
+			//if (!SSM) return;
+			
+			//functon to get money from @Arkensor & @Jacob_Mango contribution of a money System
+			//MSC = moneysystemcomponent.Cast(player.FindComponent(moneysystemcomponent))
```
Comment (Thurston00) [inline src/Scripts/Game/UI/EL_BetaHud.c:112]: This is probably better off as a `// TODO` than having pseudocode here

**Comment by LuxPoseidon** [comment]
> Couple of changes

Changes made let me know what you think

**Comment by thebonbon** [comment]
Possible to center the icons with the existing ammo icon?
ui_center

File: src/Scripts/Game/UI/EL_BetaHud.c [hunk 4]
```diff
@@ -1,3 +1,96 @@
 class EL_BetaHud : SCR_InfoDisplay
 {
-}
+	private SliderWidget staminaSlider = null;
```
Comment (salutesh) [inline src/Scripts/Game/UI/EL_BetaHud.c]: There is no need to set every method to NULL here as they are NULL by default. Other then that good code.

Comment (LuxPoseidon) [reply src/Scripts/Game/UI/EL_BetaHud.c]: This is changed in new version. See [Update EL_BetaHud.c](https://github.com/EveronLife/EveronLife/pull/78/commits/a34ad014984b170a691f25116b622116c9f63f56)

Additionally I am uploading a change to bring the UI icons lower a bit, it is hard to line up the icons as the ammo one isnt there all the time, but I think Its better now.
Newest update is out [Updated Layout](https://github.com/EveronLife/EveronLife/pull/78/commits/c18300860d78788e743688412432c52903b8119e)

**Comment by LuxPoseidon** [comment]
> Possible to center the icons with the existing ammo icon? ui_center

DONE!

File: src/Scripts/Game/UI/EL_BetaHud.c [hunk 5]
```diff
@@ -1,3 +1,103 @@
...
+			
+			//TODO: Get Survival Stats Component
+			
+			//TOFO: Get Money Stats Component
+		}; 
+		
+		OnHealthChange(m_DMC.GetHealth());
+		OnStaminaChange(m_PlayerController.GetStamina());
+		//TODO: Get info from Money and Survival Stats Components
+	}
```
Comment (Arkensor) [inline src/Scripts/Game/UI/EL_BetaHud.c:102]: We should have a null check to only access those two member variables if they are loaded.

```suggestion
	override event void UpdateValues(IEntity owner, float timeSlice)
	{
		if (!m_PlayerController)
		{
			IEntity player = SCR_PlayerController.GetLocalControlledEntity();
			if (!player) return;
		
			m_PlayerController = SCR_CharacterControllerComponent.Cast(player.FindComponent(SCR_CharacterControllerComponent));
			if (!m_PlayerController) return;
			
			m_DMC = DamageManagerComponent.Cast(player.FindComponent(DamageManagerComponent));
			if (!m_DMC) return;
			
			//TODO: Get Survival Stats Component
			
			//TOFO: Get Money Stats Component
		}; 
		
		if(m_DMC) OnHealthChange(m_DMC.GetHealth());
		if(m_PlayerController) OnStaminaChange(m_PlayerController.GetStamina());
		//TODO: Get info from Money and Survival Stats Components
	}
```

File: src/UI/Layouts/HUD/Beta/Beta.layout [hunk 6]
```diff
@@ -4,15 +4,15 @@ FrameWidgetClass {
...
+    PositionY 10.97
+    OffsetTop 10.97
     SizeX 0
-    OffsetRight 117.403
+    OffsetRight -32.198
     SizeY 0
-    OffsetBottom 42.51
+    OffsetBottom -10.97
     SizeToContent 1
    }
```
Comment (Arkensor) [inline src/UI/Layouts/HUD/Beta/Beta.layout:17]: The URL watermark widget seems to be gone now. Please move it back :)

Comment (LuxPoseidon) [reply src/UI/Layouts/HUD/Beta/Beta.layout:17]: It isn't it is in the top left, do you need / want it bottom center?

File: src/Scripts/Game/UI/EL_BetaHud.c (hunk 5 above)
Comment (LuxPoseidon) [reply src/Scripts/Game/UI/EL_BetaHud.c:102]: if (!m_PlayerController) return;
if (!m_DMC) return;

Does this, am I wrong in thinking that? The return breaks out of UpdateValues if any of them return null.

Comment (LuxPoseidon) [reply src/Scripts/Game/UI/EL_BetaHud.c:102]: I will say though there is a typo in my second TODO. WOOOOPS

**Review by Arkensor — APPROVED** [review]

**Pull request merged**